actors/evm/src/interpreter/stack.rs

```diff
@@ -10,78 +12,110 @@ const INITIAL_STACK_SIZE: usize = 32;
-        let pos = self.d - i - 1;
-        self.sk.get_unchecked(pos)
+    pub fn pop_many<const S: usize>(&mut self) -> Result<&[U256; S], StatusCode> {
+        if self.len() < S {
+            return Err(StatusCode::StackUnderflow);
+        }
+        let new_len = self.len() - S;
+        unsafe {
+            // This is safe because:
+            //
+            // 1. U256 isn't drop.
+            // 2. The borrow will end before we can do anything else.
+            //
+            // It's faster than copying these elements multiple times.
+            self.stack.set_len(new_len);
+            Ok(&*(self.stack.as_ptr().add(new_len) as *const [U256; S]))
+        }
+    }
+
+    #[inline(always)]
+    pub fn ensure_one(&self) -> Result<(), StatusCode> {
+        if self.stack.len() >= STACK_SIZE {
+            Err(StatusCode::StackOverflow)
+        } else {
+            Ok(())
+        }
     }
 
     #[inline]
-    pub unsafe fn get_mut(&mut self, i: usize) -> &mut U256 {
-        let pos = self.d - i - 1;
-        self.sk.get_unchecked_mut(pos)
+    pub fn dup(&mut self, i: usize) -> Result<(), StatusCode> {
+        let len = self.stack.len();
+        if len >= STACK_SIZE {
+            Err(StatusCode::StackOverflow)
+        } else if i > len {
+            Err(StatusCode::StackUnderflow)
+        } else {
+            unsafe {
+                // This is safe because we're careful not to alias. We're _basically_ implementing
+                // "emplace", because rust still doesn't have it.
+                //
+                // Yes, this is faster than a get/push.
+                self.stack.reserve(1);
+                *self.stack.as_mut_ptr().add(len) = *self.stack.get_unchecked(len - i);
+                self.stack.set_len(len + 1);
+            }
+            Ok(())
+        }
     }
 
     #[inline]
-    pub unsafe fn push(&mut self, v: U256) {
-        *self.sk.get_unchecked_mut(self.d) = v;
-        self.d += 1;
+    pub fn swap_top(&mut self, i: usize) -> Result<(), StatusCode> {
+        let len = self.stack.len();
+        if len < i {
+            return Err(StatusCode::StackUnderflow);
+        }
+        self.stack.swap(len - i - 1, len - 1);
+        Ok(())
     }
 
     #[inline]
-    pub unsafe fn pop(&mut self) -> U256 {
-        self.d -= 1;
-        *self.sk.get_unchecked(self.d)
+    pub fn pop(&mut self) -> Result<U256, StatusCode> {
+        self.stack.pop().ok_or(StatusCode::StackUnderflow)
     }
 
     #[inline]
-    pub unsafe fn swap_top(&mut self, i: usize) {
-        let top = self.d - 1;
-        let pos = self.d - i - 1;
-        let tmp = *self.sk.get_unchecked(top);
-        *self.sk.get_unchecked_mut(top) = *self.sk.get_unchecked(pos);
-        *self.sk.get_unchecked_mut(pos) = tmp;
+    pub fn drop(&mut self) -> Result<(), StatusCode> {
+        if self.stack.pop().is_some() {
+            Ok(())
+        } else {
+            Err(StatusCode::StackUnderflow)
+        }
```

Comment:
```suggestion
    pub fn drop(&mut self) -> Result<(), StatusCode> {
        self.stack.pop().ok_or(StatusCode::StackUnderflow).map(|_| ())
```

Reply:
Tried that, but it's slower.

Reply:
odd.

Reply:
it is not odd.

Please dont occlude branches in perf critical code.